osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs

```diff
@@ -54,6 +54,8 @@ public void Update(Playfield playfield)
 
         public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
         {
+            drawableRuleset.ResumeOverlay = null;
```

Comment:
This just cannot. I'm not sure even where to start with how wrong this is. You're setting a property to `null` without expiring, disposing, or anything. I don't even know how this would possibly work.

Reply:
> I don't even know how this would possibly work.

Well, it works because `ResumeOverlay` being `null` is handled accordingly in `DrawableRuleset`.

Anyway, would expiring the `ResumeOverlay` instead and checking `ResumeOverlay.IsAlive` here in `RequestResume()` be an acceptable solution?
https://github.com/ppy/osu/blob/8b47af6503b5a93e98940d31b96d13b1858361f5/osu.Game/Rulesets/UI/DrawableRuleset.cs#L231-L241

Reply:
I don't think that's a good direction.

I'd see this working something like:

```diff
diff --git a/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs b/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
index 6772cfe0be..ad15177898 100644
--- a/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
@@ -23,7 +23,9 @@ public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfie
         public override ModType Type => ModType.Automation;
         public override LocalisableString Description => @"Automatic cursor movement - just follow the rhythm.";
         public override double ScoreMultiplier => 0.1;
-        public override Type[] IncompatibleMods => new[] { typeof(OsuModSpunOut), typeof(ModRelax), typeof(ModFailCondition), typeof(ModNoFail), typeof(ModAutoplay), typeof(OsuModMagnetised), typeof(OsuModRepel) };
+
+        public override Type[] IncompatibleMods => new[]
+            { typeof(OsuModSpunOut), typeof(ModRelax), typeof(ModFailCondition), typeof(ModNoFail), typeof(ModAutoplay), typeof(OsuModMagnetised), typeof(OsuModRepel) };
 
         public bool PerformFail() => false;
 
@@ -60,6 +62,8 @@ public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset
 
             // Generate the replay frames the cursor should follow
             replayFrames = new OsuAutoGenerator(drawableRuleset.Beatmap, drawableRuleset.Mods).Generate().Frames.Cast<OsuReplayFrame>().ToList();
+
+            drawableRuleset.UseResumeOverlay = false;
         }
     }
 }
diff --git a/osu.Game/Rulesets/UI/DrawableRuleset.cs b/osu.Game/Rulesets/UI/DrawableRuleset.cs
index 71b452c309..d5a4ce069d 100644
--- a/osu.Game/Rulesets/UI/DrawableRuleset.cs
+++ b/osu.Game/Rulesets/UI/DrawableRuleset.cs
@@ -196,6 +196,8 @@ private void load(CancellationToken? cancellationToken)
 
             if ((ResumeOverlay = CreateResumeOverlay()) != null)
             {
+                UseResumeOverlay = true;
+
                 AddInternal(CreateInputManager()
                     .WithChild(CreatePlayfieldAdjustmentContainer()
                         .WithChild(ResumeOverlay)));
@@ -230,7 +232,7 @@ private void loadObjects(CancellationToken cancellationToken)
 
         public override void RequestResume(Action continueResume)
         {
-            if (ResumeOverlay != null && (Cursor == null || (Cursor.LastFrameState == Visibility.Visible && Contains(Cursor.ActiveCursor.ScreenSpaceDrawQuad.Centre))))
+            if (ResumeOverlay != null && UseResumeOverlay && (Cursor == null || (Cursor.LastFrameState == Visibility.Visible && Contains(Cursor.ActiveCursor.ScreenSpaceDrawQuad.Centre))))
             {
                 ResumeOverlay.GameplayCursor = Cursor;
                 ResumeOverlay.ResumeAction = continueResume;
@@ -507,6 +509,12 @@ internal DrawableRuleset(Ruleset ruleset)
         /// </summary>
         public ResumeOverlay ResumeOverlay { get; protected set; }
 
+        /// <summary>
+        /// Whether the <see cref="ResumeOverlay"/> should be used to return the user's cursor position to its previous location after a pause.
+        /// </summary>
+        /// <remarks>Defaults to <c>true</c> if a ruleset returns a non-null overlay via <see cref="CreateResumeOverlay"/>.</remarks>
+        public bool UseResumeOverlay { get; set; }
+
         /// <summary>
         /// Returns first available <see cref="HitWindows"/> provided by a <see cref="HitObject"/>.
         /// </summary>

```

Let me know if you want to add a test and update to use this direct, else I can PR an alternative for you.

Reply:
I think it would be better if we did it directly in `Player` though. This way it's handled in the same way as an instant resume during a break.

```diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -82,6 +82,11 @@ public abstract partial class Player : ScreenWithBeatmapBackground, ISamplePlaybackDisabler, ILocalUserPlayInfo

         /// </summary>
         protected virtual bool PauseOnFocusLost => true;

+        /// <summary>
+        /// Always resumes instantly without showing a <see cref="ResumeOverlay"/>.
+        /// </summary>
+        public bool AlwaysInstantResume { get; set; }
+
         public Action<bool> RestartRequested;

         private bool isRestarting;
@@ -978,8 +983,8 @@ public void Resume()
             IsResuming = true;
             PauseOverlay.Hide();

-            // breaks and time-based conditions may allow instant resume.
-            if (breakTracker.IsBreakTime.Value)
+            // breaks and time-based conditions as well as some mods may allow instant resume.
+            if (breakTracker.IsBreakTime.Value || AlwaysInstantResume)
                 completeResume();
             else
                 DrawableRuleset.RequestResume(completeResume);
--- a/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs
@@ -12,10 +12,11 @@
 using osu.Game.Rulesets.Osu.Objects;
 using osu.Game.Rulesets.Osu.Replays;
 using osu.Game.Rulesets.UI;
+using osu.Game.Screens.Play;

 namespace osu.Game.Rulesets.Osu.Mods
 {
-    public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfield, IApplicableToDrawableRuleset<OsuHitObject>
+    public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfield, IApplicableToDrawableRuleset<OsuHitObject>, IApplicableToPlayer
     {
         public override string Name => "Autopilot";
         public override string Acronym => "AP";
@@ -63,5 +64,10 @@ public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset
             // Generate the replay frames the cursor should follow
             replayFrames = new OsuAutoGenerator(drawableRuleset.Beatmap, drawableRuleset.Mods).Generate().Frames.Cast<OsuReplayFrame>().ToList();
         }
+
+        public void ApplyToPlayer(Player player)
+        {
+            player.AlwaysInstantResume = true;
+        }
     }
 }
```
Plus, it doesn't make this condition even longer than it already is.
https://github.com/ppy/osu/blob/8b47af6503b5a93e98940d31b96d13b1858361f5/osu.Game/Rulesets/UI/DrawableRuleset.cs#L233